core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/DefaultGroupVariableReference.java

```diff
@@ -0,0 +1,244 @@
+package ai.timefold.solver.core.impl.domain.variable.provided;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.ListVariableStateDemand;
+import ai.timefold.solver.core.impl.domain.variable.supply.SupplyManager;
+import ai.timefold.solver.core.preview.api.variable.provided.GroupVariableReference;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public final class DefaultGroupVariableReference<Solution_, Entity_, ParentValue_, Value_>
+        extends AbstractVariableReference<Solution_, Entity_, List<Value_>>
+        implements GroupVariableReference<Entity_, Value_> {
+    @NonNull
```

Comment:
Let's start using a class-level `@NullMarked` - that eliminates the need to use `@NonNull`.